modules/core/shared/src/main/scala/coursier/core/Dependency.scala

```diff
@@ -43,15 +47,73 @@ import dataclass.data
   def withPublication(name: String, `type`: Type, ext: Extension, classifier: Classifier): Dependency =
     withPublication(Publication(name, `type`, ext, classifier))
 
+  def withOptional(optional: Boolean) =  Dependency(module, version, configuration, exclusions, publication, optional, transitive)
+  def withTransitive(transitive: Boolean) =  Dependency(module, version, configuration, exclusions, publication, optional, transitive)
+
   lazy val clearExclusions: Dependency =
     withExclusions(Set.empty)
 
+  private val tuple = (this.module, this.version, this.configuration, this.exclusions, this.publication, this.optional, this.transitive)
   override lazy val hashCode: Int =
     tuple.hashCode()
+
+  override def toString: String = {
+    val b = new StringBuilder("Dependency(")
+    b.append(String.valueOf(module))
+    b.append(", ")
+    b.append(String.valueOf(version))
+    b.append(", ")
+    b.append(String.valueOf(configuration))
+    b.append(", ")
+    b.append(String.valueOf(exclusions))
+    b.append(", ")
+    b.append(String.valueOf(publication))
+    b.append(", ")
+    b.append(String.valueOf(optional))
+    b.append(", ")
+    b.append(String.valueOf(transitive))
+    b.append(")")
+    b.toString
+  }
 }
 
 object Dependency {
 
+  val memoised_cache: java.util.WeakHashMap[(Module, String, Configuration, Set[(Organization, ModuleName)], Publication, Boolean, Boolean), java.lang.ref.WeakReference[Dependency]] =
```

Comment:
Having both `WeakHashMap` and `WeakReference` seems fishy… Isn't only one of them enough?

Reply:
I can try, but I used I tried to do the same as stalagmite which has some nice tests...

Reply:
Re-reading this, I think I was wrong about it 😅

`WeakHashMap` has weak _keys_, but not weak values. Using a `WeakReference[_]` as value ensures the values are weak too.